test/functional/wallet_send.py

```diff
@@ -543,11 +543,16 @@ def run_test(self):
                 break
         psbt_in = dec["inputs"][input_idx]
         # Calculate the input weight
-        # (prevout + sequence + length of scriptSig + scriptsig + 1 byte buffer) * WITNESS_SCALE_FACTOR + num scriptWitness stack items + (length of stack item + stack item) * N stack items + 1 byte buffer
+        # (prevout + sequence + length of scriptSig + scriptsig) * WITNESS_SCALE_FACTOR + len of num scriptWitness stack items + (length of stack item + stack item) * N stack items
+        # Note that occasionally this weight estimate may be slightly larger than the real weight
```

Comment:
The reverse is also true - the estimate can be a bit smaller than the real weight because the tx that we are using to perform this size calculation could have a shorter signature. So I don't think this actually fixes the problem, it just results in a fee too low error occasionally rather than a fee too high error.

Reply:
The `assert_fee_amount` function allows the estimate to be a bit larger with a 2 vbyte buffer but not smaller.

I can think of a few options:

- Replicate the 2 vbyte allowance in the smaller direction
- As you already suggested switch the test to taproot

I'm leaning towards the latter as I don't like the idea of adding an allowance for every test that uses that assert.

As an aside I think it would be better to have two fee assertion functions, one that is precise and another that takes an allowance that can be specified in each tests. This way the intention of the test is clear, at the moment tests benefit from this allowance whether they need it or not.

@achow101 what are your thoughts?

Reply:
Actually, using taproot might result in a little bit less test coverage. IIRC the script chosen for the external utxo in this test is p2sh-segwit because it will have both a scriptSig and scriptWitness. This lets us check that coin control is applying the weight correctly for both legacy and segwit scripts. As Taproot is segwit only, we'd lose the coverage of the legacy side of things.

A different solution we could try is to check that the feerate is as we expected, rather than checking that the fee was expected. Since we are targeting an integer sat/vb feerate, the calculated feerate of the final tx should only ever be higher than the target by less than 1 sat/vb, so we could just calculate the sat/vbyte feerate, rounded down, and check that was what we expected.

Reply:
If the final feerate is only ever higher than the expected then the existing assertion should take care of that.

But I understood that the estimate can both be high or low depending on if a signature used for the estimate was smaller or the signature used in the final transaction was smaller.

We can still calculate the feerate in sats/vb but rounding up or down to the nearest integer.